sbi/inference/trainers/npe/npe_b.py

```diff
@@ -23,54 +23,80 @@ def __init__(
+            density_estimator: If it is a string, use a pre-configured network of the
+                provided type (one of nsf, maf, mdn, made). Alternatively, a function
+                that builds a custom neural network can be provided. The function will
+                be called with the first batch of simulations (theta, x), which can
+                thus be used for shape inference and potentially for z-scoring. It
+                needs to return a PyTorch `nn.Module` implementing the density
+                estimator. The density estimator needs to provide the methods
+                `.log_prob` and `.sample()`.
+            device: Training device, e.g., "cpu", "cuda" or "cuda:{0, 1, ...}".
+            logging_level: Minimum severity of messages to log. One of the strings
+                INFO, WARNING, DEBUG, ERROR and CRITICAL.
+            summary_writer: A tensorboard `SummaryWriter` to control, among others, log
+                file location (default is `<current working directory>/logs`.)
+            show_progress_bars: Whether to show a progressbar during training.
+        """
 
         kwargs = del_entries(locals(), entries=("self", "__class__"))
         super().__init__(**kwargs)
 
     def _log_prob_proposal_posterior(
-        self, theta: Tensor, x: Tensor, masks: Tensor
+        self,
+        theta: Tensor,
+        x: Tensor,
+        masks: Tensor,
+        proposal: Optional[Any],
     ) -> Tensor:
         """
         Return importance-weighted log probability (Lueckmann, Goncalves et al., 2017).
 
         Args:
             theta: Batch of parameters θ.
             x: Batch of data.
-            masks: Whether to retrain with prior loss (for each prior sample).
+            masks: Mask that is True for prior samples in the batch in order to train
+                them with prior loss.
+            proposal: Proposal distribution.
 
         Returns:
-            Log probability of proposal posterior.
+            Importance-weighted log-probability of the proposal posterior.
         """
 
-        raise NotImplementedError
-
-        batch_size = theta.shape[0]
-
-        # Evaluate posterior.
-        log_prob_posterior = self._posterior.net.log_prob(theta, x)
-        log_prob_posterior = log_prob_posterior.reshape(batch_size)
-        utils.assert_all_finite(log_prob_posterior, "posterior eval")
+        # Evaluate prior
+        utils.assert_all_finite(self._prior.log_prob(theta), "prior eval")
+        prior = torch.exp(self._prior.log_prob(theta))
 
-        # Evaluate prior.
-        log_prob_prior = self._prior.log_prob(theta).reshape(batch_size)
-        utils.assert_all_finite(log_prob_prior, "prior eval.")
+        # Evaluate proposal
+        # (as theta comes from prior and proposal from previous rounds,
+        # the last proposal is actually a mixture of the prior
+        # and of all the previous proposals with coefficients 1/round)
```

Comment:
This is only true when the number of simulations in each round is the same, which is not necessarily the case! I think you have to take the number of simulations added in this round, and calculate `prop` based on that.